File: C++/ChainOfResponsibility/ChainOfResponsibility.cpp
/*
 * Meyers Singleton with dead reference detect
 * Author: reimen
 * Data: Oct.6.2014
 *
 */
#include <iostream>
#include <cstring>
#include <cstdlib>
#include <stdexcept>
using std::cout;
using std::endl;
using std::string;
using std::runtime_error;
class Handler;
class ConcreteHandler2;
class ConcreteHandler3;
class ConcreteHandler5;
class Facade;

class Handler
{
public:
  virtual ~Handler() {}
  virtual void request(int num) = 0;
};

class ConcreteHandler5: public Handler
{
public:
  ~ConcreteHandler5() {}
  void request(int num)
  {
    if(num % 5 == 0)
    {
      cout << num << " is multiple of 5." << endl;
      cout << num << " is multiple of 30." << endl;
    }
    else
      cout << num << " is not multiple of 5." << endl;
  }
};

class ConcreteHandler3: public Handler
{
public:
  ~ConcreteHandler3() {}
  void request(int num)
  {
    if(num % 3 == 0)
    {
      cout << num << " is multiple of 3." << endl;
      Handler* next = new ConcreteHandler5();
      next->request(num);
      delete next;
    }
    else
      cout << num << " is not multiple of 3." << endl;
  }
};

class ConcreteHandler2: public Handler
{
public:
  ~ConcreteHandler2() {}
  void request(int num)
  {
    if(num % 2 == 0)
    {
      cout << num << " is multiple of 2." << endl;
      Handler* next = new ConcreteHandler3();
      next->request(num);
      delete next;
    }
    else
      cout << num << " is not multiple of 2." << endl;
  }
};

class Facade
{
public:
  void calc(int num)
  {
    Handler* next;
    next = new ConcreteHandler2;
    next->request(num);
    delete next;
  }
};
int main()
{
  Facade facade;
  facade.calc(30);
  facade.calc(36);
  facade.calc(2);

  return EXIT_SUCCESS;
}

File: Python/ChainOfResponsibility/ChainOfResponsibility.py
#! /usr/bin/python
# -*- coding: utf-8 -*-
'''
Chain Of Responsibility Pattern
Author: Kei Nakata
Data: Oct.10.2014
'''
import abc

class Handler(object):
    __metaclass__ = abc.ABCMeta

    @abc.abstractmethod
    def __init__(self):
        pass

    @abc.abstractmethod
    def request(self, num):
        pass

    @abc.abstractmethod
    def calc(self, num):
        pass

class ConcreteHandler5(Handler):
    def __init__(self):
        pass

    def request(self, num):
        if self.calc(num):
            print num, "is multiple of 30"

    def calc(self, num):
        if num % 5 == 0:
            print num, "is not multiple of 5"
            return True
        else:
            print num, "is not multiple of 30"
            return False

class ConcreteHandler3(Handler):
    def __init__(self):
        pass

    def request(self, num):
        if self.calc(num):
            self.nextHandler = ConcreteHandler5()
            self.nextHandler.request(num)


    def calc(self, num):
        if num % 3 == 0:
            print num, "is not multiple of 3"
            return True
        else:
            print num, "is not multiple of 3"
            return False

class ConcreteHandler2(Handler):
    def __init__(self):
        pass

    def request(self, num):
        if self.calc(num):
            self.nextHandler = ConcreteHandler3()
            self.nextHandler.request(num)

    def calc(self, num):
        if num % 2 == 0:
            print num, "is multiple of 2"
            return True
        else:
            print num, "is not multiple of 2"
            return False

if __name__ == '__main__':
    handle = ConcreteHandler2()
    handle.request(30)
    print
    handle.request(50)
    print
    handle.request(90)
    print
    handle.request(36)
